### ISSUE TITLE
 I implemented the Infini-attention from the paper "Leave No Context Behind: Efficient Infinite Context Transformers with Infini-attention" with the help of GPT, but I do not know C language. I hope that Infini-attention can be used in future model training.

### ISSUE THREAD

**Issue by breeef** [karpathy/llm.c#103]
`#include <cmath>
#include <cstring>

extern const int C; 
extern const int NH; 
extern const int HS; //HS = C / NH
extern const float scale;

void softmax(float* data, int length) {
    float maxval = -1e20f;
    for (int i = 0; i < length; i++) {
        if (data[i] > maxval)
            maxval = data[i];
    }

    float sum = 0.0f;
    for (int i = 0; i < length; i++) {
        data[i] = expf(data[i] - maxval);
        sum += data[i];
    }

    for (int i = 0; i < length; i++) {
        data[i] /= sum;
    }
}

// ELU + 1
float elu_plus_one(float x) {
    return (x >= 0 ? x : expf(x) - 1) + 1;
}

void infini_attention_forward(float* out, float* preatt, float* att, float* inp, int B, int T, int C, int NH, float* memory, float* zs) {
    int HS = C / NH;  
    float scale = 1.0 / sqrtf(HS);
    float* mem = memory; 
    float* norm = zs; 

    #pragma omp parallel for collapse(3)
    for (int b = 0; b < B; b++) {
        for (int t = 0; t < T; t++) {
            for (int h = 0; h < NH; h++) {
                float* query = inp + (b * T * C) + (t * C) + (h * HS);
                float* preatt_bth = preatt + (b * NH * T * T) + (h * T * T) + (t * T);
                float* att_bth = att + (b * NH * T * T) + (h * T * T) + (t * T);

                // Pass 1: softmax
                float maxval = -1e20f;
                for (int t2 = 0; t2 < T; t2++) {
                    float* key = inp + (b * T * C) + (t2 * C) + (h * HS);
                    float val = 0.0f;
                    for (int i = 0; i < HS; i++) {
                        val += query[i] * key[i];
                    }
                    val *= scale;
                    if (val > maxval) {
                        maxval = val;
                    }
                    preatt_bth[t2] = val;
                }

                softmax(preatt_bth, T);

                // Pass 2: 
                float* out_bt = out + (b * T * C) + (t * C) + (h * HS);
                memset(out_bt, 0, HS * sizeof(float));
                for (int t2 = 0; t2 < T; t2++) {
                    float* value = inp + (b * T * C) + (t2 * C) + (h * HS);
                    float weight = att_bth[t2];
                    for (int i = 0; i < HS; i++) {
                        out_bt[i] += weight * value[i];
                    }

                    // 
                    float elu_k = elu_plus_one(preatt_bth[t2]);
                    for (int i = 0; i < HS; i++) {
                        mem[(h * HS) + i] += elu_k * value[i];
                        norm[(h * HS) + i] += elu_k;
                    }
                }
            }
        }
    }
}

// 
void initialize_data(float* data, int size, float init_value = 0.0f) {
    for (int i = 0; i < size; i++) {
        data[i] = init_value;
    }
}

void initialize_model(int B, int T, int C, int NH, float*& inp, float*& out, float*& preatt, float*& att, float*& memory, float*& zs) {
    inp = new float[B * T * C];
    out = new float[B * T * C];
    preatt = new float[B * NH * T * T];
    att = new float[B * NH * T * T];
    memory = new float[NH * (C / NH)];
    zs = new float[NH * (C / NH)];

    // Initialize all to zeros or suitable default values
    initialize_data(inp, B * T * C);
    initialize_data(out, B * T * C);
    initialize_data(preatt, B * NH * T * T);
    initialize_data(att, B * NH * T * T);
    initialize_data(memory, NH * (C / NH));
    initialize_data(zs, NH * (C / NH), 1.0f);  // Normalize factors initialized to 1 to avoid division by zero
}
void cleanup(float* inp, float* out, float* preatt, float* att, float* memory, float* zs) {
    delete[] inp;
    delete[] out;
    delete[] preatt;
    delete[] att;
    delete[] memory;
    delete[] zs;
}

void fill_random(float* data, int size) {
    for (int i = 0; i < size; i++) {
        data[i] = static_cast<float>(rand()) / RAND_MAX;
    }
}

int main() {
    srand(time(nullptr));

    float *inp, *out, *preatt, *att, *memory, *zs;
    int B = 2, T = 10, C = 64, NH = 8;

    initialize_model(B, T, C, NH, inp, out, preatt, att, memory, zs);

    // Fill input with random data
    fill_random(inp, B * T * C);

    // Run the attention forward function
    infini_attention_forward(out, preatt, att, inp, B, T, C, NH, memory, zs);

    // Optionally, add code here to inspect the output, e.g., print or assert conditions

    // Cleanup resources
    cleanup(inp, out, preatt, att, memory, zs);

    return 0;
}`
